acquia/blt
# Fixed #1793: Allow custom commands via autoloader.

Fixes #1793 
--------

Changes proposed:
---------
This uses Robo's new class autoloading to allow third-party Composer packages to provide custom BLT commands.

Users that have already defined plugins will need to modify them slightly to use the new namespace / structure.

Additionally, hooks must now be defined within command files rather than as separate hook files.

Steps to verify the solution:
-----------
1. Verify that the built-in example command (`blt custom:hello`) works for new projects out of the box (i.e. check for regressions).
2. Verify that Composer packages can provide their own plugins. Do this by creating a new project and then adding this example repo as a composer dependency: https://github.com/danepowell/blt-example-plugin. Then you should be able to run `blt example:command-one` or `blt example:command-two`.

A few thoughts / to-dos:
- We should figure out how to handle users with existing plugins when they upgrade to 10.x. My feeling is that the number of people implementing these custom commands is small enough, and complexity involved with trying to update them large enough, that we can just advise folks of the change and let them manually update their custom commands. Open to other opinions though.
- Do we want to keep providing the "custom:hello" example command within BLT, or should we just point people to an external example Composer package? Should we clean up the namespace somehow?
